draft-irtf-cfrg-opaque.md

```diff
@@ -1171,93 +1157,310 @@ enc_server_info
 : Optional application-specific information to exchange during the protocol encrypted
 under key Ke2, defined below.
 
-mac
+server_mac
 : An authentication tag computed over the handshake transcript computed using Km2,
 defined below.
 
 ~~~
 struct {
-  uint8 mac[Nm];
+  uint8 client_mac[Nm];
 } KE3;
 ~~~
 
-mac
+client_mac
 : An authentication tag computed over the handshake transcript computed using
-Km3, defined below.
+Km2, defined below.
+
+### Key Schedule Functions
+
+#### Transcript Functions
+
+The OPAQUE-3DH key derivation procedures make use of the functions below, re-purposed
+from TLS 1.3 {{?RFC8446}}.
+
+~~~
+Expand-Label(Secret, Label, Context, Length) =
+    Expand(Secret, CustomLabel, Length)
+~~~
+
+Where CustomLabel is specified as:
+
+~~~
+struct {
+   uint16 length = Length;
+   opaque label<8..255> = "OPAQUE-" + Label;
+   opaque context<0..255> = Context;
+} CustomLabel;
 
-#### OPAQUE-3DH Key Schedule {#derive-3dh}
+Derive-Secret(Secret, Label, Transcript-Hash) =
+    Expand-Label(Secret, Label, Transcript-Hash, Nx)
+~~~
 
-OPAQUE-3DH requires MAC keys `server_mac_key` and `client_mac_key` and
-encryption key `handshake_encrypt_key`. Additionally, OPAQUE-3DH also
-outputs `session_key`. The schedule for computing this key material is below.
+Note that the Label parameter is not a NULL-terminated string.
+
+The OPAQUE-3DH key schedule requires a preamble, which is computed as follows.
 
 ~~~
-Extract("", IKM)
-    |
-    +-> Derive-Secret(., "HandshakeSecret", Hash(preamble)) = handshake_secret
-    |
-    +-> Derive-Secret(., "SessionKey", Hash(preamble)) = session_key
+Preamble(client_identity, ke1, server_identity, inner_ke2)
+
+Input:
+- client_identity,  the optional encoded client identity, which is set to its public key if not specified
```

Comment:
Also, for descriptions of client_identity, can we say "set to client_public_key if not specified" instead of "its public key"? Same with server_identity